### ISSUE TITLE
macOS AppNap put into idle YARP applications in background

### ISSUE THREAD

**Issue by francesco-romano** [robotology/yarp#1153]
Starting from macOS Mavericks (10.9) macOS supports a new feature (AppNap) with the aim of improving energy efficiency and battery life on notebooks.

Unfortunately, it uses some heuristics to decide which app may be put into "nap" (i.e. in a very low scheduling priority), see official documentation [here](https://developer.apple.com/library/content/documentation/Performance/Conceptual/power_efficiency_guidelines_osx/AppNap.html#//apple_ref/doc/uid/TP40013929-CH2-SW1).

As a consequence some yarp applications (e.g. `yarpdataplayer`) at some point stops sending data to their ports.

As documented in Apple APIs, AppNap may be disabled using (unfortunately only) ObjectiveC calls in [NSProcessInfo](https://developer.apple.com/reference/foundation/nsprocessinfo).

We have to test if this applies also to non UI applications, e.g. a simple main which in a run loop sends random data through an output port.

**Comment by vtikha** [comment]
:+1 @francesco-romano thanks for the details will look into it

**Comment by francesco-romano** [comment]
With @claudiofantacci we developed a simple macOS API wrapper to disable AppNap, see commit here: https://github.com/francesco-romano/yarp/commit/4edc999ec0fc01670fbc2596713a0ea5cc432d28

Of course this wrapper can be extended for other functions, but at the present state AppNap is the only "urgent" function to be implemented.

In [this commit](https://github.com/francesco-romano/yarp/commit/667e406ebe9c1ee0666a77eb821a2f57c591e5ee) we applied the new API to disable AppNap on `yarpdataplayer` and the results are:

```
makas-pro:yarp makaveli$ pmset -g assertions
2017-03-30 10:40:52 +0200 
Assertion status system-wide:
   BackgroundTask                 0
   ApplePushServiceTask           0
   UserIsActive                   1
   PreventUserIdleDisplaySleep    0
   PreventSystemSleep             0
   ExternalMedia                  0
   PreventUserIdleSystemSleep     1
   NetworkClientActive            0
Listed by owning process:
   pid 43394(yarpdataplayer): [0x0001e869000188a8] 00:23:54 PreventUserIdleSystemSleep named: "YARP requires AppNap off" 
```
so it is 24 minutes that yarpdataplayer streams data without stopping, and AppNap is not on as shown in the ActivityMonitor

![screen shot 2017-03-30 at 10 42 02](https://cloud.githubusercontent.com/assets/2189180/24495470/85a7f766-1535-11e7-93f0-544e05f3cdfb.png)


We now have to find a "proper" and clean way to use this new API. Something like RAII (at least for the destruction part, I personally don't like to disable AppNap it in the constructor). Feedback is welcome

**Comment by francesco-romano** [comment]
ping (for everybody) to remember this issue

**Comment by drdanz** [comment]
I don't think SystemInfo is the right place for the toggle function...

**Comment by claudiofantacci** [comment]
@drdanz where do you suggest this to go.

Please consider that for mac users this is a very important feature!

**Comment by claudiofantacci** [comment]
🤣 
That's fine. I'll speak with @francesco-romano asap 👍

**Comment by drdanz** [comment]
This was fixed in #1354 and it is available in YARP 2.3.72
